### ISSUE TITLE
Release cycle

### ISSUE THREAD

**Issue by BoboTiG** [gorakhargosh/watchdog#485]
@danilobellini what do you think to publish releases more often? For instance there are several bug fixes in the current unreleased version. Perhaps could you publish a 0.9.1 with those changes?

**Comment by danilobellini** [comment]
I think we shouldn't publish a 0.9.1 from the master tip as we've already dropped the old Python support.

What we can do:

- Release 0.10
- Cherry-pick or rewrite the bug fixes (and nothing else) in a new branch from the v0.9.0 tag to release a v0.9.1 version

Perhaps the best solution would be to do both. What do you think?

**Comment by BoboTiG** [comment]
I vote for a 0.10 release with a little change in setup.py to tell pip on what version of Python it can be installed. I can provide a patch for this, if you are OK with that.

I do not think we should take time to maintain 2 versions. Out  is incredibly limited and I do not see any benefit for no s(Pythons version dropped in 0.10 are very very old, if someone is still using one of those, no worry about watchdog version. Keeping with 0.9 is actually OK.)

**Comment by TheLastProject** [comment]
An 0.10 release would be great, especially since 48701f36513ffacaeb7b865b1cfc11e9ae1fd149. I recently added @Snyk to my project recently, and Watchdog is the only dependency I have that it considers to have known vulnerability: https://snyk.io/test/github/Pext/Pext?tab=issues. I'd love to keep my project's known vulnerability counter on 0 :)

**Comment by BoboTiG** [comment]
As 0.10 will be a breaking version, I think we could do it just after #507 (if we agree on the PR of course).

**Comment by BoboTiG** [comment]
@danilobellini what about doing a release in the next days?

**Comment by hkiang01** [comment]
> As 0.10 will be a breaking version, I think we could do it just after #507 (if we agree on the PR of course).

@BoboTiG since #507 is merged, any update?

**Comment by hkiang01** [comment]
> I think we shouldn't publish a 0.9.1 from the master tip as we've already dropped the old Python support.
> 
> What we can do:
> 
>     * Release 0.10
> 
>     * Cherry-pick or rewrite the bug fixes (and nothing else) in a new branch from the v0.9.0 tag to release a v0.9.1 version
> 
> 
> Perhaps the best solution would be to do both. What do you think?

If it's a breaking release, why not make the next version 1.0.0?

**Comment by hkiang01** [comment]
> I think we shouldn't publish a 0.9.1 from the master tip as we've already dropped the old Python support.
> 
> What we can do:
> 
>     * Release 0.10
> 
>     * Cherry-pick or rewrite the bug fixes (and nothing else) in a new branch from the v0.9.0 tag to release a v0.9.1 version
> 
> 
> Perhaps the best solution would be to do both. What do you think?

@danilobellini  If it's a breaking release, why not make the next version 1.0.0?

**Comment by BoboTiG** [comment]
I will take the responsability of the new release.
We will now try to follow the semver noatation and so, you are right @hkiang01, the next version will be 1.0.0.

**Comment by danilobellini** [comment]
I'm sorry, I lost track from what's going on here... if I can help in something please let me know. Just please make sure that the new version is still backwards-compatible as much as possible (apart from the optional `watchmedo` installation and stuff like that) and well tested, and that all changes are reported in the summary `changelog.rst` file.

Seeing the current change log, I agree the next version should be 1.0.0, and that we should use semantic versioning, marking the `Development Status` classifier as `5 - Production/Stable` or at least as `4 - Beta`. But, first, the tests should pass in the Windows builds, as well as the Mac OS build for Python 3.x (there's just a single `InterpreterNotFound` error being marked as skipped, and the whole suite marked as passing, though no test are being performed at all). I also think 49% is a way too small coverage for a stable release, and an unfair value (the platform-specific lines couldn't be counted in other platforms), perhaps we should add the project to something like [Coveralls](https://coveralls.io) to get a more meaningful coverage value.

**Comment by danilobellini** [comment]
Regarding the documentation, there are at least 3 open issues that should be addressed before AND after this new release: #410, #565 and #568. For the new release, we should probably see if the docs themselves are missing some new feature, keeping removed stuff, etc..

**Comment by luzpaz** [comment]
bump

**Comment by hsolbrig** [comment]
bump

**Comment by rrzaripov** [comment]
@BoboTiG @danilobellini what I think:
#410 - only what we need is that @gorakhargosh register project on https://readthedocs.org and provide us url and badge code, after this we can update README.rst. For example how documentation looks at rtd https://rrzaripov-watchdog.readthedocs.io/en/latest/

#565 will be closed along with #410

#568 we close after moving to rtd, and request to release new version.

So... lets do it😎

**Comment by BoboTiG** [comment]
This would be ideal, but there is one issue: @gorakhargosh never answer(s|ed). I hope he is well.
In the meantime, I do not know how to "bypass" the need for him to do something.

**Comment by CAM-Gerlach** [comment]
@BoboTiG Thanks for your work maintaining this important part of the Python ecosystem.
 If @gorakhargosh has moved on from the project or no longer has time for it, the responsible course of action would be to transfer maintainership (most importantly, the ability to cut a Github and PyPI release) to someone who can, i.e. one of you two. The alternative would be a maintained fork under e.g. a new org on Github, and under a modified or new name on PyPI.

Ensuring watchdog is compatible is a major blocker to both Python 3.8 compat and inclusion of other software in newer distro versions for a large volume of downstream projects, so particularly given that the change itself was implemented over a year ago (and the fact that `collections` for the abcs has been deprecated all the way back in Python 3.3) a release ASAP should be a high priority over a documentation issue (particularly hosting them on a website independent of the release package itself) that are rather minor by comparison to the package breaking with the current released Python version. And if the doc issues, etc. don't make it in to that release, you can simply cut another one (i.e. a point release if they are minor enough), if hosting documentation on an external website even requires an actual release at all.

Python 3.8 has already been released and will be the default in new Linux distros like Fedora and Arch, whereas watchdog is an important part of the dependency stack for a large number of projects including e.g. the Lektor static site generator, the Spyder IDE and many others, upon which much of my own work is based. If there's something relatively straightforward that a new contributor can help with to make the release happen, please let us know; given the aforementioned situation I'd be more than justified contributing to such to unblock the multiple projects I'm involved with that indirectly rely on it. I've set up and maintained the docs for Spyder a couple years back when we moved off Pythonhosted that was obsolete and frozen even then, so if there's a really legitimite reason that doing so should be a full-on blocker for a long-overdue release that would fix multiple security and/or critical breaking bugs that have been mentioned, I could try to help with that too.

**Comment by BoboTiG** [comment]
I heard you @CAM-Gerlach, and you are totally right. I totally forgot the fact that it does not work on 3.8 as I am using the mater branch in my projects.
Let's do a new release in the next week. It will have a small changelog that will be more complete later.

I will update the news for the upcoming version, 0.10.0. And upload the version on PyPi.

I am very short in spare time, are you OK @danilobellini, can I process the next week?

**Comment by danilobellini** [comment]
> [...] only what we need is that @gorakhargosh register project on https://readthedocs.org [...]

That had already been done. If it hadn't, anyone would be able to do that. It's at https://watchdog.readthedocs.io/ and I've just created it again at https://python-watchdog.readthedocs.io/en/latest/

> [...] and provide us url and badge code [...]

You mean the config to be used here in GitHub to add a docs pipeline CI through a webhook? Anyway, I just asked the author to add me as a maintainer of the first link. The webhook config doesn't suffice as no collaborator is able to change the project hooks here in this GitHub repository, just the project owner.

> [...], the responsible course of action would be to transfer maintainership (most importantly, the ability to cut a Github and PyPI release) to someone who can, i.e. one of you two. [...]

I'm able to push to PyPI a new release (I did that on v0.9.0).

> I am very short in spare time, are you OK @danilobellini, can I process the next week?

Yes, and so do I! I'll have some time next tuesday, and uploading a new release to PyPI is quite easy and fast. Just tell me when to upload it to PyPI (or something else, if you need help).

> [...] I totally forgot the fact that it does not work on 3.8 [...]

Actually, v0.9.0 works on Python 3.8.0 (I've just tested here on Arch Linux using the [python38 3.8.0-1](https://aur.archlinux.org/packages/python38/) AUR package):

```
py38 run-test: commands[0] | py.test
============================= test session starts ==============================
platform linux -- Python 3.8.0, pytest-5.2.2, py-1.8.0, pluggy-0.13.0
cachedir: .tox/py38/.pytest_cache
rootdir: /home/danilo/code/watchdog, inifile: tox.ini
plugins: cov-2.8.1, timeout-1.3.3
collected 66 items                                                             

tests/test_delayed_queue.py .                                            [  1%]
tests/test_emitter.py .........s...                                      [ 21%]
tests/test_inotify_buffer.py ......                                      [ 30%]
tests/test_inotify_c.py .                                                [ 31%]
tests/test_observer.py .....                                             [ 39%]
tests/test_skip_repeats_queue.py ....                                    [ 45%]
tests/test_snapshot_diff.py .......                                      [ 56%]
tests/legacy/test_watchdog_events.py .....................               [ 87%]
tests/legacy/test_watchdog_observers_api.py ......                       [ 96%]
tests/legacy/test_watchdog_observers_polling.py ..                       [100%]

=============================== warnings summary ===============================
.tox/py38/lib/python3.8/site-packages/watchdog/utils/bricks.py:175
  /home/danilo/code/watchdog/.tox/py38/lib/python3.8/site-packages/watchdog/utils/bricks.py:175: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.9 it will stop working

[...]

================== 65 passed, 1 skipped, 1 warnings in 23.75s ==================
___________________________________ summary ____________________________________
  py38: commands succeeded
  congratulations :)
```

To run that, I just edited `tox.ini` to install `pytest-timeout` in the `py38` environment in the `v0.9.0` tag version. Is there any Python 3.8.0-specific bug I'm missing? If so, I think that should be a new issue to be fixed ASAP. Trying to run the the current master (a60664426498a67fecebb659bb8ab9d8f53640c4) gives an error because of a dangling thread:

```
==================================== ERRORS ====================================
______________________ ERROR at teardown of test___init__ ______________________

    @pytest.fixture(autouse=True)
    def no_thread_leaks():
        """
        Fail on thread leak.
        We do not use pytest-threadleak because it is not reliable.
        """
    
        yield
    
        if sys.version_info < (3,):
            return
    
        main = threading.main_thread()
>       assert not [th for th in threading._dangling
                    if th is not main and th.is_alive()]
E       assert not [<PollingEmitter(Thread-70, stopped daemon 140145821878016)>]

main       = <_MainThread(MainThread, started 140145857795904)>

tests/conftest.py:38: AssertionError
```

**Comment by CAM-Gerlach** [comment]
Thanks for your detailed response.

> Is there any Python 3.8.0-specific bug I'm missing

The issue we were discussing was

> ``.tox/py38/lib/python3.8/site-packages/watchdog/utils/bricks.py:175
  /home/danilo/code/watchdog/.tox/py38/lib/python3.8/site-packages/watchdog/utils/bricks.py:175: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.9 it will stop working``

which is fixed in ``master`` well over a year ago shortly after the release of 0.9.0. In the 3.7 interpreter and docs the warning stated it will be removed in 3.8, not 3.9 but it looks like it just got bumped last minute. I hadn't been able to actually test my full stack yet since Conda builds of 3.8 for my non-Linux machines are a ways away and the Linux machines all had distros (Fedora 30, Debian Buster, Raspbian Buster, etc) that were still on 3.7 with the setup there being venvs with system Python. Ergo, cutting a release isn't quite so urgent, since we still have a year to go until 3.9 drops (though sooner would give downstreams more time to integrate and package it).

**Comment by BoboTiG** [comment]
@danilobellini about the dangling thread, it may be random. Or do you reproduce everytime?

**Comment by danilobellini** [comment]
> @danilobellini about the dangling thread, it may be random. Or do you reproduce everytime?

It's random, but the thread number (70) is always the same, which may be a memory leak issue or a issue with the test itself (`test___init__` in `test_observers_polling.py`). Given that the `_dangling` private attribute in threading might surprise us, I think we should use the `enumerate` method (which exists even in Python 2.7) or something else. I'm gonna make a PR with an alternative for that.

I tried running it several times (about 30 or 40 I think), it failed 4 times.

**Comment by BoboTiG** [comment]
I just updated the changelog with 587d11f. We are quite good for a new release before the end of the month, WDYT @danilobellini ?

I would like to include the PR #602 and then do you see more work to do before doing a release?
